incubator/hnc/pkg/controllers/hierarchy_controller_test.go

```diff
@@ -221,6 +222,164 @@ var _ = Describe("Hierarchy", func() {
 			return val
 		}).Should(Equal("0"))
 	})
+
+	It("should sync tree label with current hierarchy", func() {
```

Comment:
This test is very, very complicated (almost 200 lines). If it fails, it will take me a long time to figure out what the test was trying to do, which part of it broke, and why. For example, you change the parent of baz three times - why is the third time needed? Could you turn that into two tests? And can you factor out some of the code into smaller functions? Eg:

```
getDepth(ctx context.Context, from, to string) func() string {
  return func() string {
    ns := getNamespace(ctx, from)
    val, ok := ns.GetLabels[to+depthSuffix]
    return val
  }
}
```

You can use that to test the existence, nonexistence and value of the same label. Note that you need to return a _function_, not a value, since `Eventually()` will repeatedly try to call the function again.

Reply:
the third one is the case where baz (children from a parent) become root. Previously, i also thought that is similar concept with moving from parent a to parent b. But then decided to cover all possible case even though they have are similar. 

I will refactor the function as you suggest.

Reply:
In that case, perhaps factor it into two different testcases, ie "depth labels get updated when parent is changed" and "depth labels get updated when the parent is removed"?

Reply:
Got it. Updated as you suggested. PTAL

Reply:
lgtm, thanks!